### DISCUSSION TITLE
Should we allow choosing the number of spaces for indentation?

### DISCUSSION THREAD

**Discussion by DannyBen** [bashly-framework/bashly#516] [Ideas]
I noticed people (@pcrockett) using 4 space indentation in their own code, which creates an inconsistent script since bashly itself indents with 2 spaces.

Should there be a setting for it?

If we do this, we will probably be able to get rid of the "spaces to tabs" conversion done when users choose tab character indentation, and instead have the `String#indent` extension itself worry about the indentation character natively and at the source.

**Comment by pcrockett** [comment]
I don't have a strong opinion on this. The generated script is still readable even with the inconsistent indentation. Most of the time I'm reading the code under `src` anyway, not the code in the build artifact.

I lean toward waiting for a feature request from a user with a well-defined rationale. Unless it's extremely trivial to implement without making things more complex.

Side note: Might be worth thinking about how multi-line strings are handled. Adding ANY indentation to those can cause problems. I have learned to work around that personally.

Ex:

```
foo='
bar
'
```

gets transformed into:

```
  foo='
  bar
  '
```

which adds spaces in front of "bar", so my string has been unexpectedly changed.

But handling that also sounds really complex, so it's something I'm happy to work around. I'd actually rather not add any indentation at all.

**Reply by DannyBen** [reply]
Your multiline case can be solved with HEREDOC - which are not indented at all (#496).

And your script as is does not pass `shfmt` test (when executed with what should normally pass for a bashly generated script):

```
shfmt -ci -i 2 -d kbg
```

Also interesting - if I add configurable indentation, users can choose 0, to disable indentation..... is this good or bad, I am not sure...

**Reply by pcrockett** [reply]
Yeah I was thinking toward the end of my comment, "actually... maybe this is something that would be worthwhile..." because I would probably set it to 0 if the feature did exist.

But only if it didn't significantly impact maintainability of the project.

**Reply by DannyBen** [reply]
> But only if it didn't significantly impact maintainability of the project.

Should be trivial, and no additional technical debt. In fact, it will remove some in regards to the tab indentation.
I will think about it. If implemented, I will probably change the `tab_indent: true` setting, and create a unified setting `indentation: 2` or `indentation tabs`.

**Reply by DannyBen** [reply]
Well - I was painfully wrong....
I thought all my code uses the `indent` function, so the change could have been in just one place, but I also have views that already are indented naturally. So changing this is going to be difficult.
